### ISSUE TITLE
List triggers in menu

### ISSUE THREAD

**Issue by tiktuk** [federico-terzi/espanso#133]
It could be really useful if the menu listed all the triggers so one could check them without opening the config file.

**Comment by federico-terzi** [comment]
Hi @tiktuk,
Thanks for the feedback, do you mean a command for the espanso cli?
Something like:

```
> espanso triggers
:hello
:espanso
:date
```

Or do you mean an overlay GUI while typing?

**Comment by timorunge** [comment]
My two cents: The triggers are there to not look something up but in order to improve speed and productivity. This is - from my point of view - not the case if you have to take a look at the triggers that you've configured. 

A second point is that - if you have an enormous amount of triggers (I have `28062` in total) - it will be pretty much impossible to find what you're looking for.

**Comment by federico-terzi** [comment]
Honestly, I agree with @timorunge, as the whole point of espanso is to speed up the user.
That said, I can see the usefulness of some sort of *autocompletion list* that appears while typing, like in most code editors, to show the available triggers. Unfortunately, it's too complex to implement at the moment.

**Comment by tiktuk** [comment]
I did actually think about a menu with the triggers in the menubar menu. An autocompletion list could be super cool but I unedrstand that that would be a much bigger project.

I thought about a cli command listing a table with triggers / replacements would be a nice touch.

**Comment by federico-terzi** [comment]
> I actually just thought about a menu with the triggers in the menubar menu. An autocompletion list could be super cool but I understand that that would be a much bigger project.

I really appreciate your help, unfortunately at the moment it would be complicated to implement and I'd like to work on other more important features.

> I also thought about a cli command listing a table with triggers / replacements, would be a nice touch.

I'll try to work out something for this problem as it may be useful in some cases.

Thank you

**Comment by focusaurus** [comment]
Just a note on my workflow as I have experience with 2 basic approaches prior to discovering espanso.

- On mac with Keyboard Maestro I had things setup as a postfix trigger, so typing "em,," would detect the double-commas, parse the preceeding simple word ("em" in this case) and lookup the matching expansion based on that (in this case my email address)
- Now on linux, I'm using a menu-driven approach where to do an expansion I
  - Tap and release the shift key which is bound via xcape to run a script
  - The script lists all my snippet names into [rofi](https://github.com/davatorium/rofi) so I get a visual menu and can fuzzy-filter the matching snippet. This is a HUGE reliever of memory. All I need to remember is any portion of the snippet name and I can find it. If I do have the snippet already at hand in memory, it's basically as easy as just typing it like ":foo" espanso style but I tap SHIFT instead of typing ":"

So one approach would be:

- Bind a hotkey or matching typed trigger that does
  - espanso cli to write all snippet names to stdout
  - pipe that to rofi to choose which one
  - use the rofi output to make espanso perform that expansion

**Comment by federico-terzi** [comment]
Hey @focusaurus,

> * On mac with Keyboard Maestro I had things setup as a postfix trigger, so typing "em,," would detect the double-commas, parse the preceeding simple word ("em" in this case) and lookup the matching expansion based on that (in this case my email address)

This is pretty easy to achieve on espanso, you just need to put `,,` as a suffix in the trigger and espanso would do the rest. So instead of configuring the match with `em` you would use `em,,`.
That said, we can also think about an option to automatically apply a prefix or a suffix to all matches, so you don't have to do so manually.

> * Now on linux, I'm using a menu-driven approach where to do an expansion I
>   
>   * Tap and release the shift key which is bound via xcape to run a script
>   * The script lists all my snippet names into [rofi](https://github.com/davatorium/rofi) so I get a visual menu and can fuzzy-filter the matching snippet. This is a HUGE reliever of memory. All I need to remember is any portion of the snippet name and I can find it. If I do have the snippet already at hand in memory, it's basically as easy as just typing it like ":foo" espanso style but I tap SHIFT instead of typing ":"
> 
> So one approach would be:
> 
> * Bind a hotkey or matching typed trigger that does
>   
>   * espanso cli to write all snippet names to stdout
>   * pipe that to rofi to choose which one
>   * use the rofi output to make espanso perform that expansion

I like this idea. In this case, we would need a subcommand to output all the triggers to stdout and a subcommand that accepts a trigger as an argument and executes it. I'll think about it for a few days and I'll get back to you.

Thanks for your help :)

**Comment by focusaurus** [comment]
Yup. Of course this could be built by parsing the config file with existing yaml cli tools as well, maybe as a prototype.

**Comment by federico-terzi** [comment]
> Yup. Of course this could be built by parsing the config file with existing yaml cli tools as well, maybe as a prototype.

Sure, though this part could be implemented quite easily. It's the second part, the one actually executing the trigger, that requires a bit more planning.

**Comment by jonnywilliamson** [comment]
I absolutely love the sound of @focusaurus idea.

Expanso is amazing, only discovered it a week or so ago and it's installed everywhere now for me.

However I do find myself getting caught out trying to remember some of shortcuts already. I nearly always know 95% of the keyword...but don't nail it eg

My driver licence number. Something I'll only need every 12 months when typing out forms looking to renew my car insurance or PPS (social security) numbers. I'll have them set up for both my wife and me etc.... 

Was it

:jonlicense 
:jondrivlic
:jondriver
:amylicense
:amydriver


etc etc.

Having a quick match as I type would be **amazing**.

Really really hope something along these lines might be on the cards. Thanks so much for what you've done with expanso.

**Comment by federico-terzi** [comment]
Hey @jonnywilliamson 
Once the GUI is ready, this feature will definitely make its way into the roadmap!

Cheers :)

**Comment by tiktuk** [comment]
Just as a note, somebody actually created a menu with the triggers like I originally sugested: https://github.com/einstweilen/espanso-triggers :) . Haven't tried it yet, though.

**Comment by daoso** [comment]
It would be great if this worked for other triggers prefixes besides ":"
I use "<<" and my triggers don't show up in the list
